### PR TITLE
For #57 - Allows consumer to pass in their own data source for custom domains 🔃

### PR DESCRIPTION
**Pull request by boek**
Took an initial take on a possible solution for #57.

### CODE REVIEW

**Comment by boek** [comment]
While this change does have a low footprint, I'm not sure this approach makes a whole lot of sense. It might be better to decouple this from the implementations of the domain storage and having the responsibility of loading them from disk. That would also make the tests a bit cleaner and we could get rid of that flag that is for testing purposes only.

Not sure what your plans are for this but I could take another shot at it if you'd like :)

**Comment by pocmo** [comment]
I wonder why you want to implement your own provider. CustomDomains *is* the provider Focus uses. Instead we should probably add the functionality you need to CustomDomains (or wherever) so that we all can use it.

**Comment by csadilek** [comment]
@boek @pocmo Are we mainly trying to fix the problem described in https://github.com/mozilla-mobile/focus-android/issues/2000 or do we also want to load the domains from an alternative source?

For the latter, having a custom loader makes sense. The former I believe we should fix directly as it is useful functionality for the component itself.

Here's an idea how we could go about fixing the problem:

We should *not* remove the protocol before saving and instead store the URL as provided by the user. Basically, remove this line in Focus: https://github.com/mozilla-mobile/focus-android/blob/master/app/src/main/java/org/mozilla/focus/autocomplete/AutocompleteAddFragment.kt#L61 (We could still remove the prefixes for display purposes if that's desired).

Then make [tryToAutoComplete](https://github.com/mozilla-mobile/android-components/blob/master/components/browser/domains/src/main/java/mozilla/components/domains/DomainAutoCompleteProvider.kt#L105) a bit smarter by removing the protocol before matching (similar to how we handle www already). So, we'd still be auto-completing on the domain only, but we'd know the protocol.

Final step is to include the protocol as a parameter in the result callback and also in `AutocompleteResult` so we have access to it in `InlineAutocompleteEditText`, which when asked for the text, prepends the result's protocol to the domain.

**Comment by boek** [comment]
Totally, I'm still using CustomDomains in Focus. Just like @csadilek said I removed the line that trimmed the protocol before saving and my thought was to transform the list of domains before it gets loaded into DomainAutoComplete.

I'll update the PR after sprint planning 👍

**Comment by pocmo** [comment]
detekt is complaining about the line length here:

> MaxLineLength - [InlineAutocompleteEditText.kt] at components/ui/autocomplete/src/main/java/mozilla/components/ui/autocomplete/InlineAutocompleteEditText.kt:65:137

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 1]
```diff
@@ -52,23 +74,23 @@ class DomainAutoCompleteProviderTest {
...
-        val customDomains = listOf("gap.com", "fanfiction.com", "mobile.de")
+        val customDomains = listOf("gap.com", "www.fanfiction.com", "https://mobile.de")
 
         val provider = DomainAutoCompleteProvider()
         provider.initialize(RuntimeEnvironment.application, true, true, false)
         provider.onDomainsLoaded(domains, customDomains)
 
-        assertCompletion(provider, "f", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com")
-        assertCompletion(provider, "fa", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com")
+        assertCompletion(provider, "f", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com", hasWww = true)
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 2]
```diff
@@ -52,23 +74,23 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "fa", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com", hasWww = true)
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 3]
```diff
@@ -52,23 +74,23 @@ class DomainAutoCompleteProviderTest {
...
         assertCompletion(provider, "fac", AutocompleteSource.DEFAULT_LIST, domains.size, "facebook.com")
 
         assertCompletion(provider, "g", AutocompleteSource.CUSTOM_LIST, customDomains.size, "gap.com")
         assertCompletion(provider, "go", AutocompleteSource.DEFAULT_LIST, domains.size, "google.com")
         assertCompletion(provider, "ga", AutocompleteSource.CUSTOM_LIST, customDomains.size, "gap.com")
 
-        assertCompletion(provider, "m", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de")
-        assertCompletion(provider, "mo", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de")
-        assertCompletion(provider, "mob", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de")
+        assertCompletion(provider, "m", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de", protocol = "https://")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 4]
```diff
@@ -52,23 +74,23 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "mo", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de", protocol = "https://")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 5]
```diff
@@ -52,23 +74,23 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "mob", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de", protocol = "https://")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

**Comment by pocmo** [comment]
I need to setup sputnik to use our existing detekt config. Will re-run after updating sputnik!

**Comment by pocmo** [comment]
> I need to setup sputnik to use our existing detekt config. Will re-run after updating sputnik!

Argh. I can't setup sputnik to use a config file from the repository for detekt - although that's possible for other tools..

**Comment by pocmo** [comment]
> Argh. I can't setup sputnik to use a config file from the repository for detekt - although that's possible for other tools..

Filed an issue for that:
https://github.com/TouK/sputnik-ci/issues/19

Let's ignore the sputnik violations here as long as it passes on taskcluster (There seem to be some ktlint issues that need to be fixed for that).

**Comment by boek** [comment]
@pocmo Ah, interesting the first change worked for detektcheck but not ktlint. Learned some new Kotlin styles 🙌

File: components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt [hunk 6]
```diff
@@ -24,8 +25,32 @@ class DomainAutoCompleteProvider {
         const val CUSTOM_LIST = "custom"
     }
 
-    private var customDomains = emptyList<String>()
-    private var shippedDomains = emptyList<String>()
+    data class Domain(val protocol: String, val hasWww: Boolean, val host: String) {
```
Comment (pocmo) [inline components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt]: To enforce the usage of create() - would it make sense to make the constructor private or internal (if used in testing)?

File: components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt [hunk 7]
```diff
@@ -44,24 +69,26 @@ class DomainAutoCompleteProvider {
...
 
         if (useCustomDomains) {
-            val autocomplete = tryToAutocomplete(searchText, customDomains)
-            if (autocomplete != null) {
+            val result = tryToAutocomplete(searchText, customDomains)
+            if (result != null) {
+                val (autocomplete, domain) = result
                 val resultText = getResultText(rawText, autocomplete)
-                resultCallback(resultText, AutocompleteSource.CUSTOM_LIST, customDomains.size)
+                resultCallback(resultText, domain, AutocompleteSource.CUSTOM_LIST, customDomains.size)
```
Comment (csadilek) [inline components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt]: Should we pass a String to the resultCallback instead e.g. domain.toString() which could return the protocol + (www) + host? The advantages is that the caller (Focus) doesn't have to assemble the string.

File: components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt (hunk 6 above)
Comment (boek) [reply components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt]: @pocmo I had that thought too, but I changed it at the last minute to have the functionality for the AutocompleteListFragment. There is probably (definitely) a cleaner way to do this.

File: components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt (hunk 7 above)
Comment (boek) [reply components/browser/domains/src/main/java/mozilla/components/browser/domains/DomainAutoCompleteProvider.kt]: @csadilek yes, change incoming.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 8]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
 
-        assertCompletion(provider, "m", AutocompleteSource.DEFAULT_LIST, domains.size, "mozilla.org")
-        assertCompletion(provider, "www", AutocompleteSource.DEFAULT_LIST, domains.size, "www.mozilla.org")
-        assertCompletion(provider, "www.face", AutocompleteSource.DEFAULT_LIST, domains.size, "www.facebook.com")
-        assertCompletion(provider, "MOZ", AutocompleteSource.DEFAULT_LIST, domains.size, "MOZilla.org")
-        assertCompletion(provider, "www.GOO", AutocompleteSource.DEFAULT_LIST, domains.size, "www.GOOgle.com")
-        assertCompletion(provider, "WWW.GOOGLE.", AutocompleteSource.DEFAULT_LIST, domains.size, "WWW.GOOGLE.com")
-        assertCompletion(provider, "www.facebook.com", AutocompleteSource.DEFAULT_LIST, domains.size, "www.facebook.com")
-        assertCompletion(provider, "facebook.com", AutocompleteSource.DEFAULT_LIST, domains.size, "facebook.com")
+        assertCompletion(provider, "m", AutocompleteSource.DEFAULT_LIST, domains.size, "mozilla.org", "http://mozilla.org")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 9]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "www", AutocompleteSource.DEFAULT_LIST, domains.size, "www.mozilla.org", "http://mozilla.org")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 10]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "www.face", AutocompleteSource.DEFAULT_LIST, domains.size, "www.facebook.com", "http://facebook.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 11]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "MOZ", AutocompleteSource.DEFAULT_LIST, domains.size, "MOZilla.org", "http://mozilla.org")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 12]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "www.GOO", AutocompleteSource.DEFAULT_LIST, domains.size, "www.GOOgle.com", "http://google.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 13]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "WWW.GOOGLE.", AutocompleteSource.DEFAULT_LIST, domains.size, "WWW.GOOGLE.com", "http://google.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 14]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "www.facebook.com", AutocompleteSource.DEFAULT_LIST, domains.size, "www.facebook.com", "http://facebook.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 15]
```diff
@@ -36,14 +66,14 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "facebook.com", AutocompleteSource.DEFAULT_LIST, domains.size, "facebook.com", "http://facebook.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 16]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
+        val customDomains = listOf("gap.com", "www.fanfiction.com", "https://mobile.de")
 
         val provider = DomainAutoCompleteProvider()
         provider.initialize(RuntimeEnvironment.application, true, true, false)
         provider.onDomainsLoaded(domains, customDomains)
 
-        assertCompletion(provider, "f", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com")
-        assertCompletion(provider, "fa", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com")
-        assertCompletion(provider, "fac", AutocompleteSource.DEFAULT_LIST, domains.size, "facebook.com")
+        assertCompletion(provider, "f", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com", "http://www.fanfiction.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 17]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "fa", AutocompleteSource.CUSTOM_LIST, customDomains.size, "fanfiction.com", "http://www.fanfiction.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 18]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "fac", AutocompleteSource.DEFAULT_LIST, domains.size, "facebook.com", "http://facebook.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 19]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
 
-        assertCompletion(provider, "g", AutocompleteSource.CUSTOM_LIST, customDomains.size, "gap.com")
-        assertCompletion(provider, "go", AutocompleteSource.DEFAULT_LIST, domains.size, "google.com")
-        assertCompletion(provider, "ga", AutocompleteSource.CUSTOM_LIST, customDomains.size, "gap.com")
+        assertCompletion(provider, "g", AutocompleteSource.CUSTOM_LIST, customDomains.size, "gap.com", "http://gap.com")
+        assertCompletion(provider, "go", AutocompleteSource.DEFAULT_LIST, domains.size, "google.com", "http://google.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 20]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "ga", AutocompleteSource.CUSTOM_LIST, customDomains.size, "gap.com", "http://gap.com")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 21]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
 
-        assertCompletion(provider, "m", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de")
-        assertCompletion(provider, "mo", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de")
-        assertCompletion(provider, "mob", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de")
-        assertCompletion(provider, "moz", AutocompleteSource.DEFAULT_LIST, domains.size, "mozilla.org")
+        assertCompletion(provider, "m", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de", "https://mobile.de")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 22]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "mo", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de", "https://mobile.de")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 23]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "mob", AutocompleteSource.CUSTOM_LIST, customDomains.size, "mobile.de", "https://mobile.de")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

File: components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt [hunk 24]
```diff
@@ -52,24 +82,24 @@ class DomainAutoCompleteProviderTest {
...
+        assertCompletion(provider, "moz", AutocompleteSource.DEFAULT_LIST, domains.size, "mozilla.org", "http://mozilla.org")
```
Comment (sputnikci) [inline components/browser/domains/src/test/java/mozilla/components/browser/domains/DomainAutoCompleteProviderTest.kt]: \[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.

**Comment by csadilek** [comment]
@pocmo if you agree, we could release 0.4 tomorrow?

**Comment by pocmo** [comment]
@pocmo if you agree, we could release 0.4 tomorrow?

👍

**Pull request merged**